patroni/config.py

```diff
@@ -151,10 +159,87 @@ def _safe_copy_dynamic_configuration(self, dynamic_configuration):
-            elif name in config:
+            elif name in config:  # only variables present in __DEFAULT_CONFIG allowed to be overriden from DCS
                 config[name] = int(value)
         return config
 
+    @staticmethod
+    def _build_environment_configuration():
+        ret = defaultdict(dict)
+
+        def _popenv(name):
+            return os.environ.pop('PATRONI_' + name.upper(), None)
+
+        for param in ('name', 'namespace', 'scope'):
+            value = _popenv(param)
+            if value:
+                ret[param] = value
+
+        def _set_section_values(section, params):
+            for param in params:
+                value = _popenv(section + '_' + param)
+                if value:
+                    ret[section][param] = value
+
+        _set_section_values('restapi', ['listen', 'connect_address', 'certfile', 'keyfile'])
+        _set_section_values('postgresql', ['listen', 'connect_address', 'data_dir', 'pgpass'])
+
+        def _get_auth(name):
+            ret = {}
+            for param in ('username', 'password'):
+                value = _popenv(name + '_' + param)
+                if value:
+                    ret[param] = value
+            return len(ret) == 2 and ret or None
+
+        restapi_auth = _get_auth('restapi')
+        if restapi_auth:
+            ret['restapi']['authentication'] = restapi_auth
+
+        authentication = {}
+        for user_type in ('replication', 'superuser'):
+            entry = _get_auth(user_type)
+            if entry:
+                authentication[user_type] = entry
+
+        if authentication:
+            ret['postgresql']['authentication'] = authentication
+
+        users = {}
+
+        def _parse_list(value):
+            if not (value.strip().startswith('-') or '[' in value):
+                value = '[{0}]'.format(value)
+            try:
+                return yaml.safe_load(value)
+            except Exception:
+                return None
+
+        for param in list(os.environ.keys()):
+            if param.startswith('PATRONI_'):
+                name, suffix = (param[8:].rsplit('_', 1) + [''])[:2]
+                if name and suffix:
+                    # PATRONI_(ETCD|CONSUL|ZOOKEEPER|...)_HOSTS?
+                    if suffix in ('HOST', 'HOSTS') and '_' not in name:
+                        value = os.environ.pop(param)
+                        value = value if suffix == 'HOST' else value and _parse_list(value)
+                        if value:
+                            ret[name.lower()][suffix.lower()] = value
+                    # PATRONI_<username>_PASSWORD=<password>, PATRONI_<username>_OPTIONS=<option1,option2,...>
+                    # CREATE USER "<username>" WITH <OPTIONS> PASSWORD '<password>'
+                    elif suffix == 'PASSWORD':
+                        password = os.environ.pop(param)
+                        if password:
+                            users[name] = {'password': password}
```

Comment:
>If there are no quotes in the name - we should also lowercase it when reading the configuration >from the environment variable.
>> Only from environment variables? Behavior is not consistent with the config file....

You are right, we should enforce the same rule uniformly for all sources of the Patroni configuration.

>>If there are quotes - we should pass it unchanged
>Usually nobody does it. It requires to write something like:
>
>FOO='"bar"'
>what it strange and ugly... The same applies to the yaml file. If the value is quoted, it would be >treated as a string and you'll never see quotes there. Unless you will not do the same trick as >with shell.

Quoted user names are rarely used as well, so clunkiness of expressing them in the YAML is not really an issue. The idea is to lowercase all user names by default, unless the name contains quotes, in which case it should be passed verbatim to the "CREATE USER". We should also eliminate mandatory quoting from the "CREATE USER/ALTER ROLE" statement arguments.

> And I believe this is great! It automatically eliminates a lot of problems :)

And adds an inconsistency behavior, since it is possible to define  certain user names only in the configuration file, but not when getting the names from the environment variables. Besides, if we introduce a Patroni-specific limit on user names - shouldn't we document it?